sys/pm_callouts.c

```diff
@@ -885,8 +786,232 @@ void classifyAll(portmaster_packet_info* packetInfo,
+        handle = injectv6_handle;
+    } else {
+        handle = inject_handle;
+    }
+
+    // Interpret layer data as netbuffer list and check if it's a looping packet.
+    // Packets created/injected by us will loop back to us.
+    nbl = (PNET_BUFFER_LIST) layerData;
+    injection_state = FwpsQueryPacketInjectionState(handle, nbl, NULL);
+    if (injection_state == FWPS_PACKET_INJECTED_BY_SELF ||
+        injection_state == FWPS_PACKET_PREVIOUSLY_INJECTED_BY_SELF) {
+        classifyOut->actionType = FWP_ACTION_PERMIT;
+        INFO("packet was in loop, injection_state= %d ", injection_state);
+        return;
+    }
+
+    // Permit fragmented packets.
+    // But of course not the first one, we are checking that one!
+    if (FWPS_IS_METADATA_FIELD_PRESENT(inMetaValues, FWPS_METADATA_FIELD_FRAGMENT_DATA) &&
+        inMetaValues->fragmentMetadata.fragmentOffset != 0) {
+        classifyOut->actionType = FWP_ACTION_PERMIT;
+        INFO("Permitting fragmented packet: %s", print_packet_info(packetInfo));
+        return;
+    }
+
+    // get header size
+    if (FWPS_IS_METADATA_FIELD_PRESENT(inMetaValues, FWPS_METADATA_FIELD_IP_HEADER_SIZE)) {
+        ipHeaderSize = inMetaValues->ipHeaderSize;
+    } else {
+        ERR("AAAAAA!!!!");
+        classifyOut->actionType = FWP_ACTION_BLOCK;
+        return;
+    }
+    if (ipHeaderSize == 0) {
+        ERR("inMetaValues reports an ipHeaderSize of 0");
+        classifyOut->actionType = FWP_ACTION_BLOCK;
         return;
     }
+
+    // Get first netbuffer from list.
+    nb = NET_BUFFER_LIST_FIRST_NB(nbl);
+
+    // Check if this is a single netbuffer.
+    if (NET_BUFFER_NEXT_NB(nb) == NULL) {
+        FWP_ACTION_TYPE action;
+
+        // Handle netbuffer list with only one netbuffer.
+        action = classifySingle(packetInfo, verdictCache, verdictCacheLock, inMetaValues, nb, ipHeaderSize);
+        switch (action) {
+        case FWP_ACTION_PERMIT:
+            // Permit packet.
+            classifyOut->actionType = FWP_ACTION_PERMIT;
+            return;
+
+        case FWP_ACTION_BLOCK:
+            // Drop packet.
+            // TODO: Add ability to block packets, ie. respond with informational ICMP packet.
+            classifyOut->actionType = FWP_ACTION_BLOCK;
+            return;
+
+        case FWP_ACTION_NONE:
+            // Packet was fully handled by classifySingle, so we need to steal it.
+            // Block and absorb the packet.
+            // Source: https://docs.microsoft.com/en-us/windows-hardware/drivers/network/types-of-callouts
+            classifyOut->actionType = FWP_ACTION_BLOCK;
+            classifyOut->flags|= FWPS_CLASSIFY_OUT_FLAG_ABSORB;   // Set Absorb Bit to 1
+            classifyOut->rights&= ~FWPS_RIGHT_ACTION_WRITE;       // Set Write Bit to 0
+            return;
+
+        default:
+            // Unexpected value, block the packet.
+            classifyOut->actionType = FWP_ACTION_BLOCK;
+            return;
+
+        }
+    }
```

Comment:
We are not expected to ever leave the `switch` for inbound data without returning from `classifyMultiple`. Better make sure we don't forget to `return` if we add a new switch case and apply the default outside of the switch block (i.e. at the end of the if-block)